src/cinder/linux/GstPlayer.cpp

```diff
@@ -20,1298 +17,1147 @@
+    }
 	
-	return checkStateChange( stateChangeResult, targetState );
+    return checkStateChange( stateChangeResult );
 }
 
-bool GstPlayer::checkStateChange( GstStateChangeReturn stateChangeResult, const GstState & target )
+bool GstPlayer::checkStateChange( GstStateChangeReturn stateChangeResult )
 {
-	switch( stateChangeResult ) {
-		case GST_STATE_CHANGE_FAILURE:
-			g_warning( "**** Failed to change state.. ****" );
-			return false;
-			
-		case GST_STATE_CHANGE_SUCCESS:
-			g_print( "Pipeline state change SUCCESS from : %s to %s\n", gst_element_state_get_name( mGstData.mCurrentState ), gst_element_state_get_name ( target ) );
-			
-			mGstData.mTargetState = mGstData.mCurrentState = target;
-			
-			return true;;
-			
-		case GST_STATE_CHANGE_ASYNC:
-			mGstData.mTargetState = target;
-			mGstData.mCurrentState = getCurrentState();
-			mGstData.mAsyncStateChangePending = true;
-			g_print( "Pipeline state change ASYNC from : %s to %s\n", gst_element_state_get_name( mGstData.mCurrentState ), gst_element_state_get_name ( target ) );
-			
-			return true;;
-			
-		case GST_STATE_CHANGE_NO_PREROLL:
-			mGstData.mCurrentState = getCurrentState();
-			mGstData.mTargetState = target;
-			mGstData.mIsLive = true;
-			return true;
-			
-		default:
-			return false;
-	}
+    switch( stateChangeResult ) {
+        case GST_STATE_CHANGE_FAILURE: {
+	    g_warning( "** Pipeline failed to change state.. **" );
+            return false;	
+        }
+        case GST_STATE_CHANGE_SUCCESS: {
+            g_print( "Pipeline state changed SUCCESSFULLY from : %s to %s\n", gst_element_state_get_name( mGstData.currentState ), gst_element_state_get_name ( mGstData.targetState ) );
+            // Our target state is now the current one also.
+            mGstData.updateState( mGstData.targetState );
+            return true;
+        }	
+        case GST_STATE_CHANGE_ASYNC: {
+            g_print( "Pipeline state change will happen ASYNC from : %s to %s\n", gst_element_state_get_name( mGstData.currentState ), gst_element_state_get_name ( mGstData.targetState ) );
+            return true;
+        }
+        case GST_STATE_CHANGE_NO_PREROLL: {
+            mGstData.isLive = true;
+            return true;
+        }
+        default:
+            return false;
+    }
 }
 
-GstData& GstPlayer::getGstData()
+void GstPlayer::createTextureFromMemory()
 {
-	return mGstData;
+    mMutex.lock();
+    std::swap( mFrontVBuffer, mBackVBuffer );
+    mMutex.unlock();
+
+    if( mFrontVBuffer ) {
+        mVideoTexture = ci::gl::Texture::create( mFrontVBuffer, GL_RGBA, width(), height() );
+        if( mVideoTexture ) mVideoTexture->setTopDown();
+    }
 }
 
-void GstPlayer::createTextureFromMemory()
+void GstPlayer::createTextureFromID()
 {
-	mMutex.lock();
-	std::swap( mFrontVBuffer, mBackVBuffer );
-	mMutex.unlock();
+#if defined( CINDER_GST_HAS_GL )
+    mMutex.lock();
+    std::swap( mCurrentBuffer, mNewBuffer );
+    mMutex.unlock();

```

Comment:
I don't fully understand the thread syncing logic necessary here, but wouldn't the unlock need to be around the `getTextureID()` as well? Not really sure how this could be called from more than one thread, though.